build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1670,228 @@ message WaitExecutionRequest {
+      // without extensions) must be added to the observed-input digest.
+      // See comments on
+      // [ObservedPathSet][build.bazel.remote.execution.v2.ObservedPathSet]
+      // for how to generate a directory hash.
+      DIRECTORY_SEARCH = 2;
+
+      // 'Path' is a file probed for existence, but not read, by the action
+      // and that was present during execution. If the file is not present or
+      // is instead a directory, this pathset does not match the filesystem.
+      // Otherwise the hash of the string constant "F" must be
+      // added to the observed-input digest.
+      EXISTING_FILE_PROBE = 3;
+
+      // `Path` is a directory probed for existence, but not enumerated, by
+      // the action and that was present during execution. If the directory is
+      // not present, or is instead a file, this pathset does not match the
+      // filesystem. Otherwise the hash of the string constant "D" must be
+      // added to the observed-input digest.
+      EXISTING_DIRECTORY_PROBE = 4;
+
+      // `Path` is a file or directory probed for existence by the action
+      // that was not present during execution. If path is present as a file
+      // or directory, this pathset does not match the filesystem. Otherwise
+      // the hash of the string constant "P" must be added to the
+      // observed-input digest.
+      ABSENT_FILE_OR_DIRECTORY_PROBE = 5;
+    }
+
+    // File or directory path.
+    string Path = 1;
+
+    // Defines the meaning of `Path` and how to compare to the local
+    // filesystem.
+    OpKind Kind = 2;
+
+    // For entries of `OpKind.DIRECTORY_ENUMERATION`, the enumeration pattern
+    // used. When null/empty/missing the search pattern was "*" meaning the
+    // full directory.
+    string DirectorySearchPattern = 3;
+  }
+
+  // Observed path set entries. In order to ensure consistent hashing of the
+  // same `ObservedPathSet`, the entries MUST be sorted lexicographically by
+  // Path code point (or, equivalently, by UTF-8 bytes).
+  repeated Entry Entries = 1;
+
+  // For directory comparisons in Entries marked with `OpKind.DIRECTORY_SEARCH`,
+  // this defines the filenames, with extensions removed, that were read by the
+  // action. These are used as the filter for generating a directory hash to
+  // add to the observed-input digest.
+  //
+  // Design point: Extensions are removed from storage and comparisons because
+  // of search path implementations in tools that search for a varied list of
+  // extensions (e.g. .h, .hpp, no extension in C++; .dll, .exe in .NET
+  // assembly search).
+  repeated string ObservedAccessedFileNames = 2;
+}
+
+// Represents a two-phase cache pathset, referenced or inlined, with an
+// observed-input digest and a reference to build outputs to retrieve from the
+// cache if the observed-input digest calculated from the client's filesystem
+// matches.
+message ObservedPathSetAndInputDigest {
```

Comment:
A shorter name, more descriptive of its use, would most likely be better.

Reply:
We had trouble finding a good name... but PotentialActionMatch seems a lot better.